distributed/tests/test_steal.py

```diff
@@ -1182,17 +1182,20 @@ async def test_reschedule_concurrent_requests_deadlock(c, s, *workers):
     await ev.set()
     await c.gather(futs1)
 
-    # If this turns out to be overly flaky, the following may be relaxed or
-    # removed. The point of this test is to not deadlock but verifying expected
-    # state is still a nice thing
-
-    # Either the last request goes through or both have been rejected since the
-    # computation was already done by the time the request comes in. This is
-    # unfortunately not stable.
-    if victim_ts.who_has != {wsC}:
-        msgs = steal.story(victim_ts)
-        assert len(msgs) == 2
-        assert all(msg[0] == "already-aborted" for msg in msgs), msgs
+    assert victim_ts.who_has != {wsC}
+    msgs = steal.story(victim_ts)
+    msgs = [msg[:-1] for msg in msgs]  # Remove random IDs
+
+    # There are two possible outcomes
+    expect1 = [
+        ("stale-response", victim_key, "executing", wsA.address),
+        ("already-computing", victim_key, "executing", wsB.address, wsC.address),
+    ]
+    expect2 = [
+        ("already-computing", victim_key, "executing", wsB.address, wsC.address),
+        ("already-aborted", victim_key, "executing", wsA.address),
+    ]
+    assert msgs in (expect1, expect2)
```

Comment:
Before this PR, ``victim_ts.who_has != {wsC}`` was *never* true. After my change, it is *always* true. I don't think I fully understand what happened here, and if this is OK or it warrants further investigation.

- CC @fjetter 
- XREF https://github.com/dask/distributed/issues/5370

Reply:
I encountered a third possibility with `("already-aborted", victim_key, "memory", wsA.address)`

Reply:
We should move the event setting and fut gathering to the bottom. the steal responses are already logged at this point in time such that we can assert them. We only set the event and collect the futures to allow the threadpool to properly shut down but the test actually doesn't require them to finish

Reply:
This is an interesting change, indeed. It is expected.

Before, the key was _sometimes_ in state `ready`, other times in state `executing` when the steal request (B->C) reaches the workers. I don't believe that on `main` this condition is never true, rather it is unlikely.
Whether or not the task was in state ready or executing would depend on whether or not something called `ensure_computing` before the steal request could've been handled, e.g. if any other message comes over the stream before the steal request, we might've transitioned the key from ready to executing since every stream message triggers an ensure_compute.

This PR removes this ambiguity by ensuring that a task is transitioned to executing in the very same tick as it was transitioned to ready. It removes the possibility of this race condition entirely. That's great 🎉 
The ambiguity that's still in the test (expected1 vs expected2) is outside of our control since this depends on the which workers responds faster which is almost impossible to control.

we should, however, move the `ev.set()` to the bottom to remove the third psosibiltiy of the task already done computing

Reply:
If you move ev.set() to the bottom, the story is empty

Reply:
> I encountered a third possibility with `("already-aborted", victim_key, "memory", wsA.address)`

Is it
```
    expect3 = [
        ("already-computing", victim_key, "executing", wsB.address, wsC.address),
        ("already-aborted", victim_key, "memory", wsA.address),
    ]
```
or
```
    expect3 = [
        ("already-aborted", victim_key, "memory", wsA.address),
    ]
```
?